File: litedb-java/src/main/java/com/litedb/relational/TableSchema.java
package com.litedb.relational;

import java.util.ArrayList;
import java.util.List;

/**
 * TableSchema — a table's name and its typed columns.
 *
 * The PRIMARY KEY is the first column (a Phase 1 convention; an explicit PRIMARY KEY clause
 * is a later addition). Rows are stored under the key "&lt;table&gt;/&lt;primaryKeyValue&gt;".
 */
public final class TableSchema {
    public final String name;
    public final List<Column> columns;

    public TableSchema(String name, List<Column> columns) {
        if (columns.isEmpty()) {
            throw new IllegalArgumentException("table '" + name + "' needs at least one column");
        }
        this.name = name;
        this.columns = columns;
    }

    /** Convention: the first column is the primary key. */
    public Column primaryKey() {
        return columns.get(0);
    }

    public int columnIndex(String columnName) {
        for (int i = 0; i < columns.size(); i++) {
            if (columns.get(i).name.equals(columnName)) return i;
        }
        return -1;
    }

    public List<String> columnNames() {
        List<String> names = new ArrayList<>();
        for (Column c : columns) names.add(c.name);
        return names;
    }

    // ---- catalog serialization (via RowCodec) -----------------------------

    public String serialize() {
        List<String> fields = new ArrayList<>();
        fields.add(name);
        for (Column c : columns) {
            fields.add(c.name);
            fields.add(c.type);
        }
        return RowCodec.encode(fields);
    }

    public static TableSchema deserialize(String s) {
        List<String> f = RowCodec.decode(s);
        String table = f.get(0);
        List<Column> cols = new ArrayList<>();
        for (int i = 1; i + 1 < f.size(); i += 2) {
            cols.add(new Column(f.get(i), f.get(i + 1)));
        }
        return new TableSchema(table, cols);
    }

    @Override
    public String toString() {
        return name + columns + " PK=" + primaryKey().name;
    }
}

File: litedb-python/table_schema.py
"""table_schema.py — a table's name and typed columns (first column is the primary key)."""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import row_codec        # type: ignore
from column import Column   # type: ignore


class TableSchema:
    def __init__(self, name: str, columns):
        if not columns:
            raise ValueError(f"table {name!r} needs at least one column")
        self.name = name
        self.columns = columns

    def primary_key(self) -> Column:
        return self.columns[0]   # convention: first column

    def column_index(self, name: str) -> int:
        for i, c in enumerate(self.columns):
            if c.name == name:
                return i
        return -1

    def column_type(self, name: str):
        i = self.column_index(name)
        return None if i < 0 else self.columns[i].type

    def column_names(self):
        return [c.name for c in self.columns]

    def serialize(self) -> str:
        fields = [self.name]
        for c in self.columns:
            fields += [c.name, c.type]
        return row_codec.encode(fields)

    @staticmethod
    def deserialize(s: str) -> "TableSchema":
        f = row_codec.decode(s)
        cols = [Column(f[i], f[i + 1]) for i in range(1, len(f), 2)]
        return TableSchema(f[0], cols)

    def __repr__(self):
        return f"{self.name}{self.columns} PK={self.primary_key().name}"
